src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -515,6 +591,28 @@ public void AddSourceFile(string fullPath, SourceCodeKind sourceCodeKind = Sourc
             _sourceFiles.AddFile(fullPath, sourceCodeKind, folders);
         }
 
+        public async Task AddSourceFileAsync(string fullPath, SourceCodeKind sourceCodeKind = SourceCodeKind.Regular, ImmutableArray<string> folders = default)
+        {
+            var gate = new SemaphoreSlim(0);
+            void Handler(object sender, string path)
+            {
+                if (path == fullPath)
+                {
+                    gate.Release();
+                }
+            }
+
+            OnFileAdded += Handler;
+            OnFileRemoved += Handler;
+            var documentId = _sourceFiles.AddFileWithAction(fullPath, sourceCodeKind, folders, path => NotifyFileAdded(path));
+            await gate.WaitAsync().ConfigureAwait(false);
+            OnFileAdded -= Handler;
+            OnFileRemoved -= Handler;
+            gate.Dispose();
```

Comment:
all of this needs documentation.  there needs to be some explanation of hte threading/exception handling guarantees here.